File: apps/agent-service/app/schemas/kb.py
"""Schema response cho KB Inspector graph endpoints (agent-service /kb/*).

Read-only: list/paginate entity + ego-graph 1-hop. Backend proxy nguyên các response
này (xem backend-additions-plan.md §2.2/2.3).
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class EntityListItem(BaseModel):
    name: str
    norm_name: str
    type: str | None = None
    description_count: int
    source_chunk_count: int


class EntityListResponse(BaseModel):
    items: list[EntityListItem]
    total: int
    limit: int
    offset: int


class EntityEdge(BaseModel):
    source_name: str | None = None
    source_norm: str | None = None
    target_name: str | None = None
    target_norm: str | None = None
    keyword: str | None = None
    description: str = ""
    source_chunk_ids: list[str] = Field(default_factory=list)


class EntityNeighbor(BaseModel):
    name: str | None = None
    norm_name: str | None = None
    type: str | None = None


class EntityDetail(BaseModel):
    name: str
    norm_name: str
    type: str | None = None
    descriptions: list[str] = Field(default_factory=list)
    source_chunk_ids: list[str] = Field(default_factory=list)
    neighbors: list[EntityNeighbor] = Field(default_factory=list)
    edges: list[EntityEdge] = Field(default_factory=list)

File: apps/frontend/src/types/kb.ts
/** Types KB Inspector (Module 5). Mirror apps/backend/app/schemas/inspect.py +
 *  apps/agent-service/app/schemas/kb.py. */

export interface EventRef {
  event_id: string;
  label: string;
  time_start?: string | null;
  time_end?: string | null;
  confidence: string;
}

export interface ChunkListItem {
  chunk_id: string;
  heading_path: string[];
  source_file?: string | null;
  start_line?: number | null;
  end_line?: number | null;
  preview: string;
}

export interface ChunkListResponse {
  items: ChunkListItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface ChunkDetail {
  chunk_id: string;
  text: string;
  metadata: Record<string, unknown>;
  heading_path: string[];
  referencing_events: EventRef[];
}

export interface EventListItem {
  event_id: string;
  label: string;
  time_start?: string | null;
  time_end?: string | null;
  locations: string[];
  confidence: string;
}

export interface EventListResponse {
  items: EventListItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface EventDetail {
  event_id: string;
  label: string;
  summary: string;
  time_start?: string | null;
  time_end?: string | null;
  locations: string[];
  confidence: string;
  parent_event_norm?: string | null;
  source_chunk_ids: string[];
}

export interface EntityListItem {
  name: string;
  norm_name: string;
  type?: string | null;
  description_count: number;
  source_chunk_count: number;
}

export interface EntityListResponse {
  items: EntityListItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface EntityNeighbor {
  name?: string | null;
  norm_name?: string | null;
  type?: string | null;
}

export interface EntityEdge {
  source_name?: string | null;
  source_norm?: string | null;
  target_name?: string | null;
  target_norm?: string | null;
  keyword?: string | null;
  description: string;
  source_chunk_ids: string[];
}

export interface EntityDetail {
  name: string;
  norm_name: string;
  type?: string | null;
  descriptions: string[];
  source_chunk_ids: string[];
  neighbors: EntityNeighbor[];
  edges: EntityEdge[];
}
